src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+    # TODO: Add code sample when preprocessing is clearer, and encodec is added.
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPast,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+    _no_split_modules = ["BarkBlock"]
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.semantic_model = BarkCausalModule(config.semantic_config)
+        self.coarse_acoustics_model = BarkCausalModule(config.coarse_acoustics_config)
+        self.fine_acoustics_model = BarkFineAcousticsModule(config.fine_acoustics_config)
+
+        # TODO: for now, it is not integrated into the code
+        self.codec_model = EncodecModel.encodec_model_24khz()
+        self.codec_model.set_target_bandwidth(6.0)
+
+        self.config = config
+
+    def preprocess_histories_before_coarse(self, history_prompt, max_coarse_history, semantic_to_coarse_ratio, batch_size):
+        if history_prompt is not None:
+            x_semantic_history = np.repeat(history_prompt["semantic_prompt"][None], batch_size, axis = 0)  # TODO: already used before
+            x_coarse_history = history_prompt["coarse_prompt"]
+
+            x_coarse_history = (
+                _flatten_codebooks(x_coarse_history, self.config.codebook_size) + self.config.semantic_vocab_size
+            )
+            x_coarse_history = np.repeat(x_coarse_history[None], batch_size, axis = 0)
+            # e.g: after SEMANTIC_VOCAB_SIZE (10000), 1024 tokens dedicated to first codebook, 1024 next tokens dedicated to second codebook.
+
+            max_semantic_history = int(np.floor(max_coarse_history / semantic_to_coarse_ratio))
+            # trim histories correctly
+            n_semantic_hist_provided = np.min(
+                [
+                    max_semantic_history,
+                    x_semantic_history.shape[1] - x_semantic_history.shape[1] % 2,
+                    int(np.floor(x_coarse_history.shape[1] / semantic_to_coarse_ratio)),
+                ]
+            )
+
+            n_coarse_hist_provided = int(round(n_semantic_hist_provided * semantic_to_coarse_ratio))
+
+            x_semantic_history = x_semantic_history[:,-n_semantic_hist_provided:].astype(np.int32)
+            x_coarse_history = x_coarse_history[:, -n_coarse_hist_provided:].astype(np.int32)
+            # bit of a hack for time alignment (sounds better) - from Bark original implementation
+            x_coarse_history = x_coarse_history[:,:-2]
+
+        else:
+            # shape: (batch_size, 0)
+            x_semantic_history = np.array([[]]*batch_size, dtype=np.int32)
+            x_coarse_history = np.array([[]]*batch_size, dtype=np.int32)
+
+        x_semantic_history = torch.from_numpy(x_semantic_history).to(self.device)
+        x_coarse_history = torch.from_numpy(x_coarse_history).to(self.device)
+
+        return x_semantic_history, x_coarse_history
+
+    def generate_text_semantic(
```

Comment:
Doesn't it make sense for this method to belong to the text model? If we loaded the text model standalone, how would we generate currently?

Reply:
Well you would have to:
- tokenize your text
- offset the ids by config.text_encoding_offset
- add an optional semantic voice preset (Bark adds a padded array if it doesn't exist)
- concat last semantic_infer_token to the input_ids

Reply:
I think all of this still works if this method belongs to the text model rather than the composite model. We'll do all of the same pre-/post-processing, just under the text model rather than the composite one

* The tokenization happens with the `tokenizer`, not the model -> nothing changes here
* The `text_encoding_offset` parameter can go in the config no problem, we'll do all of the same token id pre-processing so this will work in the text model as well
* Same as above
* We'll do all of the post-processing as we are currently doing, so this will work in the text model as well

Reply:
Are we keeping these methods under the full model then? They've not been put under the respective sub-models